MovieQuiz/Services/QuestionFactory.swift

```diff
@@ -47,18 +50,82 @@ class QuestionFactory: QuestionFactoryProtocol {
                 text: "Рейтинг этого фильма больше чем 6?",
                 correctAnswer: false)
     ]
-    
-    init(delegate: QuestionFactoryDelegate?) {
+    */
+    init(delegate: QuestionFactoryDelegate?, moviesLoader: MoviesLoading) {
         self.delegate = delegate
+        self.moviesLoader = moviesLoader
     }
     
+    func loadData() {
+        moviesLoader?.loadMovies { result in
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                switch result {
+                case .success(let mostPopularMovies):
+                    self.movies = mostPopularMovies.items// сохраняем фильм в нашу новую переменную
+                    self.delegate?.didLoadDataFromServer()// сообщаем, что данные загрузились
+                case .failure(let error):
+                    self.delegate?.didFailToLoadData(with: error)// сообщаем об ошибке нашему
+                }
+            }
+        }
+    }
+    
+    
     func requestNextQuestion() {
-        guard let index = (0..<questions.count).randomElement() else {
-            delegate?.didReceiveNextQuestion(question: nil)
-            return
+        DispatchQueue.global().async { [weak self] in
+            guard let self = self else { return }
+            let index = (0..<self.movies.count).randomElement() ?? 0
+            
+            guard let movie = self.movies[safe: index] else { return }
+            
+            var imageData = Data()
+           
+            do {
+                imageData = try Data(contentsOf: movie.resizedImageURL)
+            } catch {
+                print("Failed to load image")
+                //возвращаемся в главный поток, сетевые данные не удалось получить, работа с ними окончена
+                DispatchQueue.main.async { [weak self] in
+                    guard let self = self else { return }
```

Comment:
👍🏻 

в новой версии свифта можно еще так 
```suggestion
                    guard let self else { return }
```